driver/main.c

```diff
@@ -157,169 +219,447 @@ PetscErrorCode SetConstantRainfall(PetscReal rain_rate, PetscInt ncells, PetscRe
+
+  data->dtime_in_hour = 1.0;  // assume an hourly dataset
+  data->ndata_file    = 1;
+
+  // load the PETSc Vec
+  PetscViewer viewer;
+  PetscCall(VecCreate(PETSC_COMM_SELF, &data->data_vec));
+  PetscCall(PetscViewerBinaryOpen(PETSC_COMM_SELF, data->file, FILE_MODE_READ, &viewer));
+  PetscCall(VecLoad(data->data_vec, viewer));
+  PetscCall(PetscViewerDestroy(&viewer));
+
+  PetscInt size;
+  PetscCall(VecGetSize(data->data_vec, &size));
+  PetscCall(VecGetArray(data->data_vec, &data->data_ptr));
+
+  // set n and s
+  data->ndata  = data->data_ptr[0];
+  data->stride = data->data_ptr[1];
+
+  // check the length of the PETSc Vec is consistent with n and s values
+  PetscCheck((size - 2) / data->stride == data->ndata, PETSC_COMM_WORLD, PETSC_ERR_USER,
+             "The length (=%d) of loaded Vec does is not consistent with first (N = %d) and second (stride = %d) in the Vec", size, data->ndata,
+             data->stride);
+
+  PetscFunctionReturn(PETSC_SUCCESS);
+}
+
+/// @brief Saves out the mapping of a dataset to RDycore as a binary Vec
+/// @param filename [in] Name of the map file
+/// @param ncells   [in] Number of RDycore's elements
+/// @param d2m      [in] Index of data-to-mesh mapping
+/// @param dxc      [in] x coordinate of the data
+/// @param dyc      [in] y coordinate of the data
+/// @param mxc      [in] x coordinate of RDycore's elements
+/// @param myc      [in] y coordinate of RDycore's elements
+/// @return PETSC_SUCESS on success
+static PetscErrorCode WriteMappingForDebugging(char *filename, PetscInt ncells, PetscInt *d2m, PetscReal *dxc, PetscReal *dyc, PetscReal *mxc,
+                                               PetscReal *myc) {
+  PetscFunctionBegin;
+
+  PetscMPIInt rank;
+  MPI_Comm_rank(PETSC_COMM_WORLD, &rank);
+
+  PetscInt stride = 6;
+  Vec      vec;
+  PetscCall(VecCreateSeq(PETSC_COMM_SELF, ncells * stride, &vec));
+
+  // fill the Vec
+  PetscScalar *vec_p;
+  PetscCall(VecGetArray(vec, &vec_p));
+  for (PetscInt i = 0; i < ncells; i++) {
+    vec_p[i * stride + 0] = i;
+    vec_p[i * stride + 1] = mxc[i];
+    vec_p[i * stride + 2] = myc[i];
+    vec_p[i * stride + 3] = d2m[i];
+    vec_p[i * stride + 4] = dxc[d2m[i]];
+    vec_p[i * stride + 5] = dyc[d2m[i]];
+  }
+  PetscCall(VecRestoreArray(vec, &vec_p));
+
+  // save the Vec
+  PetscViewer viewer;
+  PetscCall(PetscViewerBinaryOpen(PETSC_COMM_SELF, filename, FILE_MODE_WRITE, &viewer));
+  PetscCall(VecView(vec, viewer));
+  PetscCall(PetscViewerDestroy(&viewer));
+  PetscCall(VecDestroy(&vec));
+
+  PetscFunctionReturn(PETSC_SUCCESS);
+}
+
+/// @brief Sets up the mapping between the unstructured grid dataset cells and the
+///        local cells (for source-sink condition) or boundary edges (for boundary condition).
+/// @param data Pointer to a UnstructuredDataset struct
+/// @return PETSC_SUCESS on success
+static PetscErrorCode CreateUnstructuredDatasetMapping(UnstructuredDataset *data) {
```

Comment:
If the unstructured dataset itself is a mapping, would it be accurate to rename this function `CreateUnstructuredDataset`?